src/ray/raylet/worker_pool.cc

```diff
@@ -343,19 +349,59 @@ Status WorkerPool::RegisterWorker(const std::shared_ptr<Worker> &worker, pid_t p
     state.starting_worker_processes.erase(it);
   }
 
-  state.registered_workers.emplace(std::move(worker));
+  RAY_CHECK(worker->GetProcess().GetId() == pid);
+  state.registered_workers.insert(worker);
+
+  auto dedicated_workers_it = state.worker_pids_to_assigned_jobs.find(pid);
+  RAY_CHECK(dedicated_workers_it != state.worker_pids_to_assigned_jobs.end());
+  worker->AssignJobId(dedicated_workers_it->second);
+  // We don't call state.worker_pids_to_assigned_jobs.erase(dedicated_workers_it) here
+  // because we allow multi-workers per worker process.
+
   return Status::OK();
 }
 
-Status WorkerPool::RegisterDriver(const std::shared_ptr<Worker> &driver, int *port) {
+Status WorkerPool::RegisterDriver(const std::shared_ptr<Worker> &driver,
+                                  const JobID &job_id, int *port) {
   RAY_CHECK(!driver->GetAssignedTaskId().IsNil());
   RAY_RETURN_NOT_OK(GetNextFreePort(port));
   driver->SetAssignedPort(*port);
   auto &state = GetStateForLanguage(driver->GetLanguage());
   state.registered_drivers.insert(std::move(driver));
+  driver->AssignJobId(job_id);
+
   return Status::OK();
 }
 
+void WorkerPool::StartInitialWorkersForJob(const JobID &job_id) {
```

Comment:
So say I have N cores, and launch M jobs. Does this mean I end up with N*M workers by default? That seems likely to break existing workloads following this pattern.

Reply:
You are right. Previously a Ray cluster contains only one job, so M == 1 here. not a problem for existing workloads if we still want to run workloads in standalone clusters.

This design is for backward compatibility. 1) We want to start initial workers when the driver is registered to Raylet. 2) a lot of Python tests rely on initial workers.

Reply:
What about making "num_initial_workers_per_job" configurable? This is a cluster-level config. For power users who want to run multiple jobs per cluster, the cluster admin can set this to 0. And the job submiters don't need to care about this.

Reply:
But we still need to deal with the test cases unfortunately.

Reply:
Can we fix the Python tests? I can't imagine there are that many.

Reply:
@ericl I didn't count the test cases. I bet there are about a dozen test cases that depend on initial workers. Those test cases assert some conditions which depend on living workers, but they don't create any tasks. That's why they depend on initial workers. I can definitely fix these test cases. I've tried to fix them before. See [this PR](https://github.com/ray-project/ray/pull/7020).